src/fn_call.rs

```diff
@@ -71,8 +76,196 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
+                // This ensures that we always check the validity of the argument,
+                // even if we don't end up using it
+               
+                trace!("__rustc_start_panic: {:?}", this.frame().span);
+
+
+                // Read our 'usize' payload argument (which was made by casting
+                // a '*mut &mut dyn BoxMeUp'
+                let payload_raw = this.read_scalar(args[0])?.not_undef()?;
+
+                // Construct an ImmTy, using the precomputed layout of '*mut &mut dyn BoxMeUp'
+                let imm_ty = ImmTy::from_scalar(
+                    payload_raw,
+                    this.machine.cached_data.as_ref().unwrap().box_me_up_layout
+                );
+
+                // Convert our ImmTy to an MPlace, and read it
+                let mplace = this.ref_to_mplace(imm_ty)?;
+
+                // This is an '&mut dyn BoxMeUp'
+                let payload_dyn = this.read_immediate(mplace.into())?;
+
+                // We deliberately do this after we do some validation of the
+                // 'payload'. This should help catch some basic errors in
+                // the caller of this function, even in abort mode
+                if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Abort  {
+                    // FIXME: Actually print out the payload here
+                    return err!(MachineError("the evaluated program abort-panicked".to_string()));
+                }
+
+                // This part is tricky - we need to call BoxMeUp::box_me_up
+                // on the vtable.
+                //
+                // core::panic::BoxMeUp is declared as follows:
+                //
+                // pub unsafe trait BoxMeUp {
+                //     fn box_me_up(&mut self) -> *mut (dyn Any + Send);
+                //     fn get(&mut self) -> &(dyn Any + Send);
+                // }
+                //
+                // box_me_up is the first method in the vtable.
+                // First, we extract the vtable pointer from our fat pointer,
+                // and check its alignment
+
+                let vtable_ptr = payload_dyn.to_meta()?.expect("Expected fat pointer!").to_ptr()?; 
+                let data_ptr = payload_dyn.to_scalar_ptr()?;
+                this.memory().check_align(vtable_ptr.into(), this.tcx.data_layout.pointer_align.abi)?;
+
+                // Now, we dereference the vtable pointer.
+                let alloc = this.memory().get(vtable_ptr.alloc_id)?;
+
+                // Finally, we extract the pointer to 'box_me_up'.
+                // The vtable is layed out in memory like this:
+                //
+                //```
+                // <drop_ptr> (usize)
+                // <size> (usize)
+                // <align> (usize)
+                // <method_ptr_1> (usize)
+                // <method_ptr_2> (usize)
+                // ...
+                // <method_ptr_n> (usize)
+                //```
```

Comment:
We already have vtable generation code in `librustc_mir::interpret::traits`. Why do we need it again here?

Reply:
Oh, wait, this is *reading* the vtable?
Can you make it share code with the `terminator` implementation in the core engine that resolves virtual function calls?

Reply:
This comment still stands.